packages/@lwc/state/src/index.ts

```diff
@@ -1 +1,165 @@
+    this.computeValue();
+    super.notify();
+  }
+
+  get value() {
+    if (!this._value) {
+      this.computeValue();
+    }
+    return this._value;
+  }
+}
+
+const isUpdater = (signalOrUpdater: Signal<unknown> | ExposedUpdater) =>
+  typeof signalOrUpdater === 'function';
+
+const atom: MakeAtom = <T,>(initialValue: T) => new AtomSignal<T>(initialValue);
+
+const computed: MakeComputed = (inputSignalsObj, computer) =>
+  new ComputedSignal(inputSignalsObj, computer);
+
+const update: MakeUpdate<AtomSignal<unknown>> = <
+  AdditionalArguments extends unknown[],
+  Updater extends (signalValues: ValuesObj, ...args: AdditionalArguments) => ValuesObj,
+  SignalsObj extends Record<string, AtomSignal<unknown>>,
+  ValuesObj extends {
+    [signalName in keyof SignalsObj]?: UnwrapSignal<SignalsObj[keyof SignalsObj]>;
+  },
+>(
+  signalsToUpdate: SignalsObj,
+  updater: Updater,
+) => {
+  return (...uniqueArgs: AdditionalArguments) => {
+    const signalValues = {} as ValuesObj;
+
+    for (const [signalName, signal] of Object.entries(signalsToUpdate)) {
+      signalValues[signalName as keyof ValuesObj] = signal.value as ValuesObj[keyof ValuesObj];
+    }
+
+    const newValues = updater(signalValues, ...uniqueArgs);
+
+    for (const [atomName, newValue] of Object.entries(newValues)) {
+      signalsToUpdate[atomName][atomSetter](newValue);
+    }
+  };
+};
+
+export const defineState: DefineState = (defineStateCallback) => {
+  return (...args) => {
+    class StateManagerSignal<OuterStateShape> extends SignalBaseClass<OuterStateShape> {
+      private internalStateShape: Record<string, Signal<unknown> | ExposedUpdater>;
+      private isNotifyScheduled = false;
+      private _value: OuterStateShape;
+
+      constructor() {
+        super();
+
+        // @ts-ignore: TODO
+        const fromContext: MakeContextHook = () => {};
+
+        this.internalStateShape = defineStateCallback(atom, computed, update, fromContext)(...args);
+
+        for (const signalOrUpdater of Object.values(this.internalStateShape)) {
+          if (!isUpdater(signalOrUpdater)) {
+            // Subscribe to changes to exposed state atoms, so that the entire state manager signal
+            // "reacts" when the atoms change.
+            (signalOrUpdater as Signal<unknown>).subscribe(this.scheduledNotify.bind(this));
+          }
+        }
+      }
+
+      private computeValue() {
+        const computedValue = Object.fromEntries(
+          Object.entries(this.internalStateShape).map(([key, signalOrUpdater]) => {
+            if (isUpdater(signalOrUpdater)) {
+              return [key, signalOrUpdater];
+            }
+            return [key, signalOrUpdater.value];
+          }),
+        );
+
+        this._value = Object.freeze(computedValue) as OuterStateShape;
+      }
+
+      private scheduledNotify() {
+        if (!this.isNotifyScheduled) {
+          this.isNotifyScheduled = true;
+          queueMicrotask(() => {
```

Comment:
@divmain I added the microtask to notify the `StateManagerSignal` only once for all the notifications trickled by Atoms and Computed etc.
This might be a no-go for various reasons if we want the updated to be sync, let me know your thoughts on this.

Reply:
The idea makes sense but I think it would have an undesired side-effect if multiple state atoms were updated simultaneously in a single updater. For each of the atoms that's updated by the updater, `scheduledNotify` would fire. And therefore, for each of the atoms, a microtask would be enqueued to notify downstream consumers. Additionally, we'd get redundant invocations of `computeValue`.

I made some suggestions for lazy evaluation of `ComputedSignal`; could you try implementing something similar for `StateManagerSignal`?

Reply:
The current implementation uses a flag to add a task to microtask queue only **_once_**. I like your approach of lazy evaluation of computation but it essentially still notifies state manager multiple time for example ( 2 times ) [here](https://github.com/salesforce/lightning-labs/pull/34/files#diff-1e3baa7816e18c04f0cf40742ad28d783d90a4967220ee3fe2a37165c0359132R49)

I was trying to avoid this essentially because ultimately this could trigger multiple DOM updates, let me know if my understanding is not correct ?

The other approach I was thinking of is if we can somehow know how many notifications an _**Update**_  action would trigger but that got overcomplicated really fast.

Reply:
I'm a bit surprised that the spy function fires twice in the linked test. Can you help me understand why that is happening?

That said, I was originally thinking that it wouldn't matter how many times a consumer is notified. This is because, when LWC consumes the signal, it has its own run-once mechanism to trigger an async re-render. So if the subscriber gets called multiple times, that's ultimately okay and would be taken care of at the web framework level.

On the other hand, while enqueuing a microtask shouldn't be strictly necessary, but it doesn't seem like bad either.

So depending on the answer to why the subscriber is called twice - I do want to know why that's happening, in case there's an internal logic bug we've overlooked - perhaps we can combine `isNotifyScheduled` with the lazy computation. Those two behavioral characteristics do not conflict, I don't think.

Reply:
@divmain 
>I'm a bit surprised that the spy function fires twice in the linked test. Can you help me understand why that is happening?

The state manager subscriptions should be notified for all the state changes, since internally State Manager is subscribing to notifications for individual atom/computed changes. When I perform an update `increment`, 2 things are changing, `count` and `doubleCount` so those 2 notifications represent those.

>So if the subscriber gets called multiple times, that's ultimately okay and would be taken care of at the web framework level.

It sounds somewhat similar to the `queueMicrotask` approach if I am not mistaken. Trying to understand why do you think it's better to delegate this task to rendering framework instead of handling it here ?

> perhaps we can combine isNotifyScheduled with the lazy computation

Agreed on this!

Reply:
Thanks for the explanation, that is obvious in retrospect and makes complete sense.

My opinion on this comes from two things:

1. A desire for simplicity: since there is already a de-dupe mechanism in LWC, we don't need to implement it here.
2. Deduping, at least as originally implemented, involves adding some asynchronicity (with `queueMicrotask`) to what is otherwise synchronous.
    1. That might be entirely fine but it also might not be desirable; I'm not really sure yet.
    2. We can always make a synchronous thing asynchronous later, whereas we can't necessarily accomplish the inverse and make switch from async back to sync.

I would not be surprised if we had another PR in a short while that re-adds the `queueMicrotask` stuff. But I'd like to try out the simpler version first with all other pieces in the larger platform to see if we can make a more informed choice.